Explicit specificity

goldfirere opened the pull request:
This proposal introduces new syntax `typeRep :: forall {k} (a :: k). ...` (the braces are new) to allow a user to quantify a variable without affecting downstream users who might use visible type application.

[Rendered](https://github.com/goldfirere/ghc-proposals/blob/explicit-specificity/proposals/0000-explicit-specificity.rst)

AntC2 commented:
>  the braces here serve to make the argument "more invisible".

ROTFL

`{ }` as the cloak of more invisibility.

treeowl commented:
I think this is a grand idea. I'm still a bit uncomfortable with allowing variables to be specified when they're not explicitly quantified, as in

```haskell
p :: a -> b -> F a b  -- p @A @B is valid
q :: forall (a :: k). a -> G a  -- p @K @A is valid
```

I would generally prefer to see this effect from

```haskell
p :: forall a b. F a b
q :: forall k (a :: k). a -> G a
```

but I recognize that being liberal about what's specified is helpful when using functions that weren't specifically written for `TypeApplications` but that work just fine with it anyway, and that I will probably have my head bitten off if I propose that a variable must be quantified explicitly in order to be specified.

Ericson2314 commented:
I think this isn't completely future-compatible? This might be valid someday:

```haskell
{-# LANGUAGE NamedFieldPuns #-}
data Foo = Foo { i :: Natural }

forall (Foo { i } :: Foo) . Vec a i -> Vec a i
```

treeowl commented:
@Ericson2314 That strikes me as somewhat surprising hypothetical syntax. Why use that instead of

```haskell
forall a (foo :: Foo). foo ~ 'Foo { i } => Vec a i -> Vec a i
```

I mean, I guess you could argue that your way reads better, but it seems like unnecessary complication.

simonpj commented:
I like this proposal. Internally, GHC already has "inferred" and "specified", but it's just inaccessible to the user.  (Mind you, I always trip over nomenclature.)

The "Proposed specification" is not nearly precise enough.  Beyond syntax I think we need to say

* The new inferred variables participate in the `-XScopedTypeVariables` feature; for example, given
```
f :: forall {k} (a::k). blah
f = e
```
both `k` and `a` scope over the right hand side `e`.

* In a use of Visible Type Application (`-XTypeApplications`), the visible type arguments instantiate only the "specified" arguments, not the "inferred" ones.

We should also specify what happens in data/class declarations.

Under "Motivation", consider this:
``
f :: forall a. Int -> Proxy (a::k) -> blah
f = e
```
In today's GHC, with `-XScopedTypeVariables`, the type variable `a` scopes over `e`.   But so does `k`!!! Even though it only has occurrences (no binders) and is buried deep inside the type.  This is ghastly, but is caused by the fact that initially we had no way to bind `k`.  With `TypeInType` we now do; but only like this
```
f :: forall k a. Int -> Proxy (a::k) -> blah
-- or, stylistically better
f :: forall k (a::k). Int -> Proxy a -> blah
```
and now in a visible type application I'd have to specify the kind argument, which I might not want to.   This proposal makes it possible to bind `k` without also making it a positional argument for VTA.  Obviously more sensible.

vagarenko commented:
There still will be a way to specify kind parameter if needed? 

If I want to apply `foo :: forall {k} (a :: k). ...` to say `Nothing :: Maybe Bar`. How do I write it?
`foo @(Nothing :: Maybe Bar)`  ?

Ericson2314 commented:
@treeowl

I chose my example for a reason :). It's unclear whether `i` is bound implicitly, or normally with a named field pun.

int-index commented:
I am in favor of this proposal.

However, the overall design raises a question. We appear to have three levels of visibility for a parameter: visible, invisible, and super-invisible (inferred).

Is there no way whatsoever to specify the super-invisible parameters? Is this asking for a `@@` syntax?

int-index commented:
Perhaps we would want visibility levels?

0. visible
1. invisible, introduced with `forall a`, overriden with `@`
2. invisible, introduced with `forall {a}`, overriden with `@@`
3. invisible, introduced with `forall {{a}}`, overriden with `@@@`

etc

goldfirere commented:
@treeowl: There was some debate about whether or not to make user-written, implicitly quantified variables available for type application. We decided "yes" in the end, but this was not an obvious choice, at the time or in retrospect. I still favor this approach as it vastly enlarges the applicability of visible type application, but I won't defend the choice staunchly.

@Ericson2314: That's an interesting new syntax. Let me expand your idea a bit. In some hypothetical language that came before Haskell (cough, cough, ML), we can imagine that pattern matching must be done explicitly:

```
f :: Maybe a -> Bool
f x = case x of
  Nothing -> False
  Just _ -> True
```

But in Haskell, we can pattern-match directly on function arguments:

```
f :: Maybe a -> Bool
f Nothing = False
f (Just _) = True
```

How convenient. You seem to be proposing (in some far-off future) that we do something similar in type variable binders. Just as the difference between my snippets above is that the latter allows pattern-matching in a variable binding position, you want to do the same, but in types. It's a reasonable idea. And if we allow type patterns to appear in type variable binding positions, I agree that the proposed syntax is indeed *not* future-compatible. I'm not sure how worried to be about this, but I'm glad you brought it up. Do you have any other syntax ideas?

Regardless, I've added this concern to the proposal.

@simonpj: I've added a note to the specification in the proposal that braced variables work with `-XScopedTypeVariables`. Your second bullet about inferred vs specified is no change from current behavior, but I now remind readers of this fact in my revised proposal. For your extra motivational example, you're assuming a change to the way scoping works, which will appear in another proposal. I don't think it's necessary to bring this change into scope here.

@vagarenko: No. If a variable is *inferred*, then it's not available with visible type application. Full stop. This fact keeps type inference predictable. The idea is that inferred type variables are invented by GHC, and it's conceivable that different minor versions of GHC might do it slightly differently. If there were a way to observe the ordering of inferred variables, then programs could break. Of course, these new braced variables aren't really inferred in the same way (instantiating them *would* be predictable), but allowing an override here but not for GHC-inferred variables leads to a system where we have specified, user-written-inferred, and truly-inferred variables in the mix, and we'd be teetering toward madness.

@int-index: You've anticipated parts of my answer to @vagarenko. No, this doesn't lead to three levels of visibility: just two. They are *specified* and *inferred*. That's it! No `@@`!

goldfirere commented:
OK -- well, there are three levels of visibility. But *visible* isn't addressed in this proposal. It's addressed in #81 and doesn't interact with this proposal. In particular, `forall {k} (a :: k) ->` (using the syntax of #81) will be an error.

int-index commented:
> and it's conceivable that different minor versions of GHC might do it slightly differently.

Okay, so let us consider the visibility levels model I mentioned earlier:

0. visible
1. invisible, introduced with forall a, overriden with @
2. invisible, introduced with forall {a}, overriden with @@
3. invisible, introduced with forall {{a}}, overriden with @@@

In this model, we could also talk about infinity-level (inferred) — those variables can be introduced by GHC exclusively, and there is no possibility for an override.

After this proposal we have an asymmetry: the users can introduce the inferred variables, but cannot override them. This is not a deal breaker, but the abstraction is leaking.

goldfirere commented:
What abstraction is leaking? I see an asymmetry, but I don't see a leaky abstraction.

int-index commented:
The abstraction (to my understanding) is that the "inferred" level is an implementation detail for GHC. Now it's leaking to the user-facing language (the ability to write `forall {a}`) — yes, out of necessity, but so is the case with other abstraction leaks. And the asymmetry is more evidence that it's a leak.

(Once again, I'm in favor of the proposal, I see the need for it, just sharing an observation that bothers me a bit)

goldfirere commented:
OK -- I can see what you're getting at now, and I see why it's bothersome. But I don't think we can do better.

int-index commented:
> But I don't think we can do better.

Can we ensure that the variables introduced by the user with the `forall {k}` syntax are the same as what GHC might have inferred? So this would be a means to access inferred variables, not to introduce them.

int-index commented:
Since GHC infers these type variables, this is similar to partial type signatures, which GHC infers also.

In the motivation section, we have this example:

```
typeRep1 :: Typeable a => TypeRep a
typeRep2 :: Typeable (a :: k) => TypeRep (a :: k)
```

With `-XPartialTypeSignatures` we can rewrite it as:

```
typeRep3 :: Typeable a => TypeRep (a :: _)
```

And then introduce _named partial type signatures_, which allow access to whatever GHC infers:

```
typeRep3 :: Typeable a => TypeRep (a :: (k = _))
                                         ^
                                    this is a binding
                                    site for 'k'
```

(the syntax is hypothetical)

Now `k` is whatever GHC infers, and can be used in `-XScopedTypeVariables`. This is also more general because the inferred type need not be a variable. And it's entirely orthogonal to `-XTypeApplications`, where the abstraction barrier is preserved.

vagarenko commented:
@goldfirere 
> No. If a variable is inferred, then it's not available with visible type application.

But I didn't use visible type application for the inferred variable in my example:

>If I want to apply `foo :: forall {k} (a :: k). ...` to say `Nothing :: Maybe Bar`. How do I write it?
`foo @(Nothing :: Maybe Bar)`?

I use type application for `a` variable, not `k` and give it kind `Maybe Bar` via type annotation. I'm asking will this continue to work? It works in GHC-8.2.1:

````
>:set -fprint-explicit-kinds
>:t Proxy @'Nothing
Proxy @'Nothing :: forall {a}. Proxy (Maybe a) ('Nothing a)
>:t Proxy @('Nothing :: Maybe Int)
Proxy @('Nothing :: Maybe Int) :: Proxy (Maybe Int) ('Nothing Int)
````

goldfirere commented:
@int-index: You are suggesting replacing

```
typeRep :: forall {k} (a :: k). Typeable a => TypeRep a
```

with

```
typeRep :: forall (a :: (k = _)). Typeable a => TypeRep a
```

where the `k` would scope over the implementation of `typeRep`. However, these two type signatures mean different things: the first requires that the kind of `a` be abstract, while the second allows `a`'s kind to be inferred to be something other than a variable. Furthermore, your proposed `k = _` syntax isn't necessary: see [NamedWildCards](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#named-wildcards).

Bottom line: I don't think this is an adequate replacement for this proposal.

@vagarenko: Ah, I understand better now. Yes, that would still work. This proposal is all about how to write types to work the desired way with visible type application.

int-index commented:
> However, these two type signatures mean different things: the first requires that the kind of a be abstract, while the second allows a's kind to be inferred to be something other than a variable.

Well, that's the point (sort of). Rather than add another mechanism to introduce variables, we just use "whatever GHC infers", which might not be a variable (I don't see how it's problematic).

> Bottom line: I don't think this is an adequate replacement for this proposal.

Okay, but can you add it to alternatives? There's probably a use case where this alternative wouldn't work, but I can't come up with one.

Bj0rnen commented:
What's the argument for adding this syntax instead of allowing to declare type variables out of dependency order natively? With that, no special syntax would be needed, and you could effectively "hide" type variables just by putting them last.

It seems to me like support for that would only require a trivial amount of sugar on top of what this proposal does. At least this example made it seem that way:
```haskell
typeRep4 :: forall {k} (a :: k) k'. (k ~ k', Typeable a) => TypeRep a
```
So what if you could just write that as below?
```haskell
typeRep4 :: forall (a :: k) k. Typeable a => TypeRep a
```
Internally it could be implemented using the proposed hiding mechanism without having to expose that as a language feature. Some care has to be taken to not expose the `k'` variable that would be generated in any error messages, etc. Not that there would need to be a full-blown implementation of the proposed feature anyway.

1. Is it actually always trivial to desugar out-of-dependency-order type variables into the synatx proposed here?
2. Does pure hiding add something to the table which free reordering doesn't?

Ericson2314 commented:
@goldfirere 

> Do you have any other syntax ideas?

@Bj0rnen 

That's nice, but anti-modular. Say you are using a type alias:
```
old :: OldFoo -- OldFoo is higher ranked, so old takes type arguments
backCompat :: forall {a}. NewFoo a
```
Admittedly, this came out more contrived than I thought, but the basic idea is if some quantifiers are abstracted away, it's not easy to rearrange them.

Ericson2314 commented:
Overall, I am sort of getting the same cat-and-mouse vibes between introduction and elimination as @int-index with the `@@@...` examples. The motivation is compelling, and I can't think of another solution, but the situation overall seems...tragic.

goldfirere commented:
I'm intrigued by @Bj0rnen's idea. It can be backward-compatible: as long as a library doesn't change existing type parameters, that library can always add new ones at the end of a list. I'm not convinced by the anti-modular argument @Ericson2314 puts forward. Implementing this idea seems very difficult, unfortunately. (Currently, GHC uses the same types in Core as it does in Haskell, and @Bj0rnen's idea means we either have to stop doing this or somehow encode a transformation between them.) I'd be curious for more feedback on this idea.

goldfirere commented:
I've added @Bj0rnen's idea as an alternative, and I argue there why I think the main proposal is still the best way forward.

int-index commented:
@goldfirere What's a use case that `NamedWildCards` for a kind would not cover, but the current proposal would?

> Bottom line: I don't think this is an adequate replacement for this proposal.

goldfirere commented:
@int-index GHC will rightly accept `forall (a :: _k). a -> a`, filling in `_k` with `Type`, whereas it rightly won't accept `forall {k} (a :: k). a -> a`. I would want a user to be able to write the latter, and have GHC confirm that the kind really is a skolem variable, not a hole for some other type.

int-index commented:
>  I would want a user to be able to write the latter, and have GHC confirm that the kind really is a skolem variable, not a hole for some other type.

Why? I see no motivation in the proposal. Assume GHC infers a concrete type instead of a fresh variable. Worst-case scenario: the user gets a type-error. With the current proposal: the user gets a different type error.

Does a different type error justify adding new syntax and an abstraction leak?

goldfirere commented:
But the type error with `NamedWildCards` might conceivably appear only in a library's client code, not the library itself. Another way to put this: we can imagine a Haskell-like language that didn't have user-written type variables at all, only named wildcards. Then, you would write `id :: _a -> _a` and hope your implementation is actually parametric. (Indeed, I believe that standard ML does this.) Such a language seems inferior to what we actually have, though. Why should this smaller case be different?

I know I'm contradicting myself here, but I can't seem to understand how this is a leaky abstraction. According to the [visible type application paper](https://repository.brynmawr.edu/cgi/viewcontent.cgi?article=1001&context=compsci_pubs), inferred variables are part of the type system specification, not part of GHC's implementation. This proposal merely gives the user direct access to more of the type system.

goldfirere commented:
Bah. I've just come up with a problem with this proposal.

A [previous (accepted) proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0007-instance-foralls.rst) of mine adds a way to explicit quantify variables in places that this is otherwise impossible. That proposal misses some cases though. Consider

```
class C (a :: Proxy k) where ...
```

Where do we bind `k`? Where do we give it a type? There are no answers to these questions. The "obvious" answer looks like this:

```
class C {k :: Maybe Bool} (a :: Proxy k) where ...
```

Here, the braces make `k` go from *visible* to *specified*. In the proposal under discussion, the braces make a variable go from *specified* to *inferred*. What if we wanted a class variable to be *inferred*? I suppose we could do

```
class C2 {{k :: Maybe Bool}} (a :: Proxy k) where ...
```

Why does this matter? Class variables are prepended to method types, so the distinction between *specified* and *inferred* matters, even if we don't have [visible kind application](https://github.com/treeowl/ghc-proposals/blob/type-level-type-app/proposals/0000-type-level-type-applications.rst). But we will have visible kind application someday, and then this matters more.

We could just have a rule thus: braces increase the invisibility of a binder by 1 degree. This is consistent both with the explicit specificity proposal and my new extension in this comment. Is it confusing? And, it seems to suggest we should indeed have more than 3 degrees of visibility (although this is not required for it all to hold together).

Another possibility is to require double-braces for all *inferred* variables, regardless of whether they appear in a type declaration or in a `forall` in a type signature. (Double-braces also neatly sidesteps the `-XNamedFieldPuns` potential ambiguity that @Ericson2314 dreamed up.)

What do folks think of this? I do think we should resolve the syntax for binding invisible variables in datatype declarations before deciding on the explicit specificity proposal.

simonpj commented:
I'm confused by your "Bah" commment.  Let's see if I have it right.

First, a reminder:

* **Inferred**: the type argument cannot be specified at a call site; it is always inferred.

* **Specified**: the type argument may be specfied at a call site, via Visible Type Application, but it is not compulsory.

* **Required**: the type argument must be specified explicitly at every call site.  (This only happens in types, currently, never in terms.)

Let's assume we want to write explicit foralls as much as possible.  (In my view this should always be possible.)

For **type signatures in terms** we can write something like
```
f :: forall a. a -> blah      -- Specified
f :: forall {a}. a -> blah    -- Inferred (this proposal)
```
Now, in terms a type arguments is never visible; that is, it is never compulsory at a call site.  But your "Bah" comment is pointing out that for type/class decls we have three possibilities, and we want syntax for all three.   I though you had in mind
T :: forall k. k -> blah      -- Specified
T :: forall {k}. k -> blah    -- Inferred (this proposal)
T :: forall k -> k -> blah    -- Required
```
using the `->` after the `forall`.  I think there's a proposal about this too.

Your "Bah" comment seems to be concerned with class declarations.   You want syntax for all three situations.  If we had separate kind signatures (which is the subject of another proposal, which I can't find now) we could use the above syntax.  But I suppose that even if we did, we might want the class decl itself to specify which of hte three is needed.  One possibility would be (as you suggest):
or (as you suggest)
```
class C k     ... where ...   -- Required   C :: forall k -> k -> blah
class C {k}   ... where ...   -- Specified  C :: forall k. k -> blah
class C {{k}} ... where ...   -- Inferred   C :: forall {k}. k -> blan
```
But it is obviously horrible that the syntax of the class decl and the syntax of the the kind of the class (both given above) differ!   So you suggest an alternative:

```
class C k     ... where ...   -- Required   C :: forall k.     k -> blah
class C {k}   ... where ...   -- Specified  C :: forall {k}.   k -> blah
class C {{k}} ... where ...   -- Inferred   C :: forall {{k}}. k -> blan
```
If we were going to do this, we would not need the `forall k -> blah` syntax any more.

Another alternative would be to use a different piece of syntax, say angle-brackets:
```
class C k   ... where ...   -- Required   C :: forall k.     k -> blah
class C <k> ... where ...   -- Specified  C :: forall <k>.   k -> blah
class C {k} ... where ...   -- Inferred   C :: forall {k}. k -> blan
```
Finally, we could say that you must use an explicit kind signature to get any kind of quantification other than explicit, thus:
```
class C :: forall k -> k -> Constraint where ...   -- Required   C :: forall k -> k -> blah
class C :: forall k. k -> Constraint   where ...   -- Specified  C :: forall k.   k -> blah
class C :: forall {k}. k -> Constraint where ...   -- Inferred   C :: forall {k}. k -> blan
```
We would still want the `forall k` to scoe over the `...` part thought.

simonpj commented:
Another very attractive option is to abolish "inferred" altogether.  That is, arguments would always be either visible (compulsory) or specified (optional).  

That would kill off a whole mess of complexity.   I don't know why I didn't think of this before.  The inferred/specified distinction is solving a future problem, one we do not yet have.

goldfirere commented:
I agree with your review of the current situation, up to "So you suggest an alternative:". My suggested alternative is

```
class C k (a::k)     where ...   -- Required   C :: forall k ->   k -> Constraint
class C {k} (a::k)   where ...   -- Specified  C :: forall k.     k -> Constraint
class C {{k}} (a::k) where ...   -- Inferred   C :: forall {{k}}. k -> Constraint
```

which is uglier but backward compatible. (Yours is not, changing the meaning of `forall .`)

I'm a bit uncomfortable with the weird scoping rules that would come into play if we allow kind signatures on classes, but I could be convinced that *inferred* variables in type declarations are available only by a top-level kind signature (as proposed in #54).

We can't drop *inferred* altogether, and I'm not sure what you mean about a future problem. In terms, it's very much a present problem, as explained at length in the [Visible Type Application paper](https://repository.brynmawr.edu/cgi/viewcontent.cgi?article=1001&context=compsci_pubs). Because we don't have visible type application in types, we don't yet need this distinction in type declarations. However, we hopefully will soon: #80. And, while this proposal doesn't include syntax for such things, I would want us to design the syntax for term-level type signatures keeping in mind what might happen in the future.

On balance, perhaps the best answer is to say that you need to use a top-level kind signature (à la #54) if you want an explicit binding site for a type variable in a type declaration, sidestepping this whole debate.

simonpj commented:
> Yours is not, changing the meaning of forall .

Changing it how?  Give an example or two?

> weird scoping rules that would come into play if we allow kind signatures on classes

What weird rules?  Are they not the same as on term type-signatures?

>  In terms, it's very much a present problem

Can you be more specific about where this lengthy explanation is?   If we only had specified, not inferred, what bad things happen?

goldfirere commented:
> Changing it how? Give an example or two?

Right now, if I say `Proxy :: forall k. k -> Type`, that `k` is *specified*. That is, we write `Proxy True`. As you originally described it, `Proxy :: forall k. k -> Type` would mean that `k` is *required*, meaning we would write `Proxy Bool True`.

>>    weird scoping rules that would come into play if we allow kind signatures on classes

> What weird rules? Are they not the same as on term type-signatures?

Yes. We would have `class C :: forall k. blah where` where the `k` would scope also over the `where`. I hate this in term type signatures, and I don't wish to propagate this wonky design.

> Can you be more specific about where this lengthy explanation is? If we only had specified, not inferred, what bad things happen?

See section 3.1 of the paper. Note that the paper uses "generalized" where GHC uses "inferred", but these terms are in direct correspondence. The problem of living without inferred variables is that we lose predictability of type inference. This is precisely what was hard about visible type application in the first place.

simonpj commented:
> See section 3.1 of the paper. 

I quote "In other words, we allow visible type application to instantiate a polytype only
when that type is given by a user annotation." 

So according to the paper, we have no need for the user to specify inferred quantifiers.  If the user specifies it, the it's, well, specified.

But this proposal says "if we have compiler-introduced inferred quantifiers, then we want to display them; and then we want the programmer to write them too", which is a significant departure from the paper, and drags in quite a lot of new issues, as we see above.  So we were sold A, but now we discover that A drags in B.  If we'd known that we might have viewed A differently.

For example, we could specify the order of inferred quantifiers as left-to-right.  That has its problems, to be sure, but it feels as if we are on a slippery slope at the moment.  Maybe we should crawl back to the top.

goldfirere commented:
I don't know how we could do left-to-right ordering of inferred quantifiers, especially in the presence of constraints (which are fundamentally unordered), never mind type synonyms (which can change order), and perhaps other thwarts.

So maybe the answer is to be against this proposal. I think crawling back up to the top requires something like one of
1. Scrap visible type application
2. Abandon the principle of predictable type inference

Another possible way forward is to fix the motivators of this proposal in another manner. Fore example, we could come up with a new way to scope type variables that doesn't require an explicit mention in a type signature and could allow type variables to appear out of dependency order. These features might be good on their own, but the first will be a challenge to design and the second will be a challenge to implement. Those barriers haven't stopped us before, however...

simonpj commented:
> Abandon the principle of predictable type inference

I think that's putting it a bit strongly.  Assume left-to-right ordering, then almost always we get a predictable order of quantifiers.  The helpful summary in Fig 2 suggests that the cases were you don't are relatively rare. And if you care, add a type signature.

Here's another idea: currently VTA is used to instantiate specified variables, but there can be a mixture of specified and inferred.  But we didn't really seek an interleaved mixture of specified and inferred variables.  We were really just trying to distinguish functions that have a (fully specified) signature from functions that lack one.   So here's another possible rule: we still have inferred/specified/required, but you can only use VTA if the entire initial sequence of quantifiers is specified.

So given
```
typeRep1 :: forall a. Typeable a => TypeRep a
typeRep2 :: forall k (a::k). Typeable (a :: k) => TypeRep (a :: k)
```
you could only use VTA for `typeRep2`.

(Incidentally, I think we could reasonably restrict specified variables for just the explicitly frorall'd ones, so the order if 100% visible.)

This is a bit more restrictive than now; but it means that the user would never need to write down an inferred binder; it'd only be used as a way for the compiler to know "I can't use VTA here because it's inferred".  Which was the original intent.

goldfirere commented:
Maybe my "no more predictable type inference" was a bit strong. But it's very possible that the left-to-right order that GHC infers changes between minor versions. (Consider what happens if you have `type SwapEither b a = Either a b` somewhere.) Also, the person who cares is the library client, but the person who would have to add the type signature is the library author. Of course, a client can simply redefine a library export with a type signature, but that doesn't seem like something we want to encourage.

You've made a few other suggestions here:

1. VTA only when all variables are Specified. But then we're back to having `typeRep @_ @True` instead of the more ergonomic `typeRep @True`, and this would be a breaking change.

2. VTA only when the variables are explicitly forall'd. We considered this with the initial release of the feature, but then it means that all those libraries who don't bother to put `forall` in their type signatures lose out. This would be another breaking change.

I'm afraid none of these look better than the status quo or this proposal, to me.

simonpj commented:
Well (1) looks attractive to me; arguably it's what we should have done in the first place.  It's breaking change but only for people using VTA.  And this entire business of mixing specified and inferred gives me the shivers.  If you want visible type application, you should supply all the arguments!

If we adopted this we would not need this entire proposal; and your "Bah" (to which we have no satisfactory resolution) would be moot.  So I suppose that (1) is really an alternative proposal.

(2) was a side remark; it's entirely orthogonal.

goldfirere commented:
I still don't understand (1). Take this as the example:

```
typeRep :: Typeable a => TypeRep a
```

Under your suggestion, this would not be available for visible type application, because it's quantified over an inferred variable `k` and a specified one `a`. But now consider this:

```
f :: forall (a :: Type). TypeRep a -> ()
f _ = ()

x = f (typeRep @Int)
```

That looks like (1) would accept it. Bidirectional type inference tells us that `typeRep`'s *inferred* argument must be `Type`. Then, there are only specified arguments left (just one of them), and so visible type application becomes available. Urgh! It just seems utterly non-compositional. Also, what about `bar :: Int -> forall a. TypeRep a -> Bool`? `bar` will have an inferred argument `k`. But `bar 3` wouldn't anymore (the `k` has to have been instantiated, as it comes before the `Int`). I think this cure is worse than the disease, sorry to say.

simonpj commented:
> That looks like (1) would accept it. 

No.  `typeRep` has an inferred quantifier, so you can't use VTA.  (1) says you can't use VTA if there are any inferred quantifiers in the way.  It's simple, does not require reasoning about inference, and fully compositional.

goldfirere commented:
OK -- fair enough. But your rule still wouldn't work with the theory laid out in the paper. A critical part of Damas-Milner type inference (which GHC is still based on) is that the declarative specification can generalize *anywhere*. Visible type application doesn't change that, *as long as the generalized variables are infererd*. Under your rule, the introduction of a fresh inferred variable would indeed change the behavior of a function at a use site, by suddenly forbidding type application. I guess this is what I meant by non-compositional. Under your rule, you could have `f :: forall a. ...` and `f @Int` would type-check. Yet if you have `/\{b}. f` (where `b` is inferred)`, then `(/\{b}. f) @Int` wouldn't type-check, *even though instantiating `b` should be invisible*. Something seems very off to me here.

simonpj commented:
> could have f :: forall a. ... and f @Int would type-check. 
Yes if you mean `f :: forall (a:Type). ...`.   Not if you mean `f :: forall {k} (a:k). ...`.

> A critical part of Damas-Milner type inference 

Can you explain why it's critical?  GHC does not randomly generalise and re-instantiate.

All I'm suggesting is that if a type is specified by the programmer, then it should be *fully* specified by the programmer, not part-specfiied, part-inferred.  That makes things simpler!

goldfirere commented:
That property of HM is important because it means, essentially, that you `let`-abstract and `let`-inline willy-nilly without noticing, given that `let` does generalization. This, in turn, means that a most general type really is best: you can always replace the type of an expression with its most general type, and the program still is accepted.

However, with visible type application, we've already broken the "willy-nilly" property. If you `let`-abstract an expression with specified type variables, those variables become inferred. If you`let`-inline, you have to include the type signature as an ascription.

You've done well defeating my objections. But I'm not done yet.

It is possible to have an inferred variable come after a specified one:

```
class C b where
  meth :: b -> TypeRep a -> ()
```

GHC will conclude `meth :: forall (b :: Type). C b => forall {k} (a :: k). b -> TypeRep @k a -> ()`. The first required argument is the argument of type `b`; everything before that is invisible. This telescope of arguments includes one inferred argument, `k`. Does that mean that users will be unable to specify `b`? I suppose it does.

What irks me with this example is that the reasoning requires searching up until the first visible argument to see if there are any inferred arguments. It's non-local. Currently, we can just look at the first argument and proceed.

Even without these objections, I think that the rule makes programming with visible type application more awkward, not less. You might argue that it makes it all *simpler* at the cost of some convenience, but I don't necessarily agree with that. We'll get questions why we can't use VTA with `f :: forall a. TypeRep a -> ()`. Explaining that is no simpler than explaining some of the details around inferred variables, I think.

Saagar-A commented:
@goldfirere 

> ```Haskell
class C {k} (a::k)   where ...   -- Specified  C :: forall k.     k -> Constraint```

Wouldn't it make sense to reuse the kind application syntax for introducing specified kind variables?

```Haskell
class C @k (a ∷ k) where 
```
@simonpj 

> Another very attractive option is to abolish "inferred" altogether. That is, arguments would always be either visible (compulsory) or specified (optional).
>
>That would kill off a whole mess of complexity. I don't know why I didn't think of this before. The inferred/specified distinction is solving a future problem, one we do not yet have.

Couldn't all these problems be side stepped if GHC allowed named application of non-visible parameters?
Then, the order would not be significant. Syntax could be one of the following

```Haskell
map (a = Int) negate [1..10]
map @(a = Int) negate [1..10]
map @{ a = Int } negate [1..10]
```

goldfirere commented:
Ooh. I very much like `class C @k (a :: k)`. We could also have `class C2 @(k :: Maybe Bool) (a :: Proxy k)`. It doesn't allow us to make an inferred variable, but I'm OK saying that if you want that, use a top-level kind signature à la #54.

I (with collaborators, such as @sweirich) have considered named parameters. The problem with named parameters is it makes a seemingly-local choice exported. When I say

```
id :: a -> a
id x = x
```

I generally think of that as fully identical to

```
id :: b -> b
id x = x
```

I might have even written `forall a. a -> a` and `forall b. b -> b`. However, under named parameters, then these signatures are observably different, as the type parameters have different names. That's a bit frustrating, because I really think of `forall a. a -> a` and `forall b. b -> b` as identical constructs. I'm also not sure how the theory with named parameters would work out (corresponding to the work in the "Visible type application" paper).

goldfirere commented:
By the way, this `class C @k (a :: k)` syntax suggested by @Saagar-A is nice enough that I'm no longer as concerned as I was when writing [my "Bah" comment](https://github.com/ghc-proposals/ghc-proposals/pull/99#issuecomment-355781604). So I'm happy to consider this proposal unblocked and will update the proposal accordingly.

int-index commented:
@goldfirere I see the value of alpha-equivalence, but I don't think we are violating it here, just making it non-local.

We have that

```
f :: forall a. a -> a
g = f @(a = Int)
```

is equivalent to

```
f :: forall b. -> b
g = f @(a = Int)
```

This name is not a part of program's semantics — it does not affect runtime behavior in any way. I would be more worried about something like this:

```
data D = D deriving Show
main = print D
```

which changes its behavior when I perform a renaming:

```
data T = T deriving Show
main = print T
```

and yet we seem not to worry about this.

At the same time, with named `forall` I can always change an API to use different names:

```
f :: forall a. a -> a

f' = forall b. b -> b
f' = f @(a = b)
```

so it's very well-contained. And the ergonomics boost is immense, I think — this will act in two complementary ways:

* motivate people to provide descriptive type variable names
* document call sites, making it evident which parameters are specified

Besides, this frees us of thinking about the order of the arguments. I really think it's a huge win. There's a precedent to named arguments, they have it in OCaml, and no one seems to complain :)

goldfirere commented:
Idris also does it with named arguments. I'm curious to hear what others think, too.

However, I will say that changing to named arguments would be a significant change and would certainly require its own proposal. There would also be backward-compatibility issues to resolve with today's `-XTypeApplications`. I'm not saying we shouldn't explore this, but there are real costs here.

I suppose (though this needs to be explored further) that we could have *both* ordered arguments and named arguments.

AntC2 commented:
> ... I'm curious to hear what others think, ...
> I suppose (though this needs to be explored further) that we could have both ordered arguments and named arguments.

I can see the _prima facie_ appeal of named arguments for optional extras. As @goldfirere says that's a major change to how Haskell works.

I've used languages with _both_ named and positional (quite a lot of CLs/scripting do this because there's typically a gazillion optional args). But the naming typically is available only for _trailing_ argument positions, not _leading_ as we want for explicit specificity; and they don't support partial application. I think a combination of named, positional and partial application will be a swamp:

- What if the second arg is named `b` and you supply both two positional args and one named `b`? Does that mean use the second positional argument for position 3? (Typically scripting languages reject this as too confusing.)
- What if you partially apply to one arg: can the usage site that consumes that apply it to named `b`? Then we need to switch on this extension in the client module(?)
- Or perhaps we have a rule that a named arg can only appear in leading positions(?) Again how does that go with partial application?

goldfirere commented:
I'm worried that discussion of named arguments is drifting away from the original scope here. That said, if we only ever look at small proposals, we'll never get the potentially beneficial high level view. So, I'll continue this debate, but with caution.

One way of thinking about all this is:
1. Suppose we didn't have any way of applying a function to type arguments. Would we prefer
  a. named arguments
  b. positional ones (what we have now)
  c. or both?
2. If the answer to (1) is not (b), is it worth the cost of getting from here to there?

I agree that @AntC2 makes a compelling case against option (c). One challenge in this debate is that, having implemented and worked with (b), we can see its warts. (a)'s warts are still cloaked in mystery, but perhaps someone with experience with, e.g., Idris may have an opinion here. (Of course, Haskell tends to take type inference more seriously than other languages, so what works there might not work here. For example, Coq uses positional parameters and doesn't bother at all with specified-vs-inferred. But Coq also can't infer the type of `id x = x`.)

nomeata commented:
Not much happening here; marking as dormat for now.

goldfirere commented:
I've woken it up, and this proposal is ready to be served to the committee. Thanks, @nomeata .

To summarize the above discussion: there was some talk of abandoning the inferred/specified distinction altogether, but I think that distinction is here to stay, as explained in the visible type application paper. Conversation also veered toward having named type parameters instead of positional ones, but conversation died out there, and that idea is much bigger than this one proposal.

AntC2 commented:
I see @yav has posted some (useful) questions for clarification about how the proposal works [on the Committee mailing list](https://mail.haskell.org/pipermail/ghc-steering-committee/2018-May/000534.html).

Could I ask that such questions and their answers appear in this discussion thread. (And possibly get incorporated into the proposal text, if it's unclear/incomplete.)

I appreciate it's often difficult to separate out aspects of the discussion but as a rule of thumb: here is about the what and how; the Committee discussion should be only why and why important/worthwhile, and assuming those Committee members who post have already absorbed the proposal. (At least, that's how I understand it.)

int-index commented:
It still bothers me that, assuming we get `forall ->`, this leaves us with these variations:

* type variables that are visible by default, `forall a ->`
* type variables that are invisible by default but can be overriden, `forall a.`
* type variable that are invisible and cannot be overriden, `forall {a}`.

This means that we have a language mechanism to override visibility at call site, and then another mechanism to say "no, this variable visibility cannot be overriden". I fail to see why we stop at this point and do not introduce the next level, "yes, I *do* want this variable visibility to be overriden", ad infinitum.

This leads us to the design I outlined in https://github.com/ghc-proposals/ghc-proposals/pull/99#issuecomment-353071898

As long as we don't introduce `forall {a}.`, we have the wiggle room to retreat and consider things like named type applications, for which I argue in https://github.com/ghc-proposals/ghc-proposals/pull/99#issuecomment-360857229

I would like to avoid committing to the current suboptimal solution.

simonpj commented:
> It still bothers me that, assuming we get forall ->, this leaves us with these variations:
> 
> 1.     type variables that are visible by default, forall a ->
> 2.    type variables that are invisible by default but can be overriden, forall a.
> 3.    type variable that are invisible and cannot be overriden, forall {a}.

It seems to me that we are more or less forced into this design:

1. Some arguments are compulsory.  When we write `Maybe Int` we must write the `Int`.  We can't write `Maybe` and expect the `Int` to be inferred.
2. We want visible type application.   It's in GHC now and it's extremely useful.
3. But currently we can write type signatures that do have explicit foralls but that do not quantify, or even mention, all their kind variables.  We can't reasonably use visible type application for these because their order is so unconstrained.

I suppose that we could declare class (3) empty, by **rejecting any typethat has variables that are not explicitly mentioned**, and thus live in class (3).   So then
```
data Proxy a = Proxy
```
would not be poly-kinded -- instead I suppose it'd default to kind `Type`.  To get the poly-kinded version you'd have to write
```
data Proxy (a :: k) = Proxy
```
And I suppose that might not be a bad thing.

But what about term-level functions whose type is inferred?  (A bit like `Proxy`, but at the term level.)  We clearly still have to accept functions without a type signature.  But (in the current design) their type variables are in category (3) and we can't use visible type application.

AntC2 commented:
> But what about term-level functions whose type is inferred? (A bit like `Proxy`, but at the term level.) We clearly still have to accept functions without a type signature. But (in the current design) their type variables are in category (3) and we can't use visible type application.

I think we have a general (but probably informal) rule that if you want to use 'fancy' type features, you must put a type signature. This is in line with Hindley-Milner inference, which never _requires_ a signature.

> 2. We want visible type application. It's in GHC now and it's extremely useful.

Yes and yes. Speaking as somebody still trying to get my head round type application, I always feel nervous about the canonical sequence of arguments. Having to guess it for tokens that don't even have a signature strikes me as too hard. As it is, you have to pay attention it's not as simple as left-to-right appearance of variables in a signature.

Then you're implying a rule that seems sensible: to use type application, put a signature; if that's poly-kinded type application, that signature must mention variables for the kinds.

AntC2 commented:
> it's hard to make do with two [categories].

(Must admit I've lost the plot on all the options swirling around here; and I've just read the 'Type Variables in Patterns' paper ...)

Can't we resort to type signatures in patterns, and thereby 'reach' any type/kind, even if we can't reach it via type applications? To take your `Proxy` without explicit kind

    data Proxy a = Proxy                          -- I'm importing this, so can't change it

    f :: Proxy (a :: k) -> blah ... a ... k ...
    f (x :: (Proxy (aa :: kk))) = ... aa ... kk ...

(I've used distinct `tyvars` there for the pattern, to show there's nothing up my sleeves. Or is TVIP proposing to deprecate the `PatternTypeSignatures` bit of `ScopedTypeVariables`? I do see it's going to be confusing having so many similar-but-different ways to express typing.)

The TVIP paper points out pattern signatures can be clumsy/ugly. (Presumably this arises because the code wanting to put signatures is importing something not declared with type applications in mind.) There is a case [paper section 3.6] where it's impossible to specify a suitable pattern signature. But the paper's proposing a way to express that with type applications. It's an existential GADT with a type family; so falls well within my 'fancy' features; then there has to be a signature. (And hooray for doing away with `Proxy`.)

goldfirere commented:
I've updated this proposal to remove the ability to put braces in the tyvars in a class head, in favor of new proposal #148.

nomeata commented:
I am still undecided whether I like this approach (encoding specificity in the type) or would rather see that as separate declarations (like we do for the roles of type constructor). But I think I know what questions I have to ask to make up my mind:

Will the following work:
```
-- specificity in a higher rank position
foo1 :: (forall {a} b. a -> b -> a) -> Int
foo1 g = g @String 0 "Hi"

-- specificity overriden by type annotation
foo2 :: Int
foo2 = (const :: forall {a} b. a -> b -> a) ) @String 0 "Hi"

-- partial application
foo3 :: Int
foo3 = h @String (True, "Hi")
  where
    h = g 0
    g :: forall a. a -> (forall {b} c. (b,c) -> a)
    g x _ = x
```

If the answer is “yes”, then I agree with putting the specificity into the types, because it behaves like types.

If the answer is “no”, ans specificity is really a property of “the $n$-th type argument of a concrete, bound function”, then I am inclined to advocate for a separate declaration (like `roles`). This would also avoid stealing the `{…}` syntax in types, which may be valuable in the future, and obviate the need for different  printing modes in `:t` because specificity could be reported in `:info name`.

I could not answer these questions clearly from the spec, which may mean that the spec could be refined to do so. (Or I missed the crucial bit if its there, which is possible :-))

NB: Coq has both forms of declarations (`{a}` in types, `Arguments` command), but does *not* support it in higher rank contexts, which I found a unsatisfying experience.

goldfirere commented:
In response to https://github.com/ghc-proposals/ghc-proposals/pull/99#issuecomment-401060976

* `foo1` is accepted.
* `foo2` is accepted.
* `foo3` is rejected. In no change from the status quo, all `let`-bound definitions are instantiated and regeneralized en route to determining their types. Thus, `h` gets the type `forall {a} {b} {c}. Num a => (b, c) -> a` and is not available for visible type application.

In other words, these really are types, not some other construct.

nomeata commented:
Thanks!

ChShersh commented:
I appreciate this proposal as well. Just today I really needed the feature. I will describe my use-case below.

## Problem description 

I have data type parameterized by type variable polymorphic in kind:

```haskell
data TypeRepMap (f :: k -> Type) ...
```

And I would like to write the following code:

```haskell
lookup :: forall a f . TypeRepMap f -> Maybe (f a)
lookup = ... some implementation here ...

member :: forall a f . TypeRep f -> Bool
member trMap = case lookup @a trMap of
    Nothing -> False
    Just _ -> True
```

Unfortunately, this code doesn't compile with the following error:

```
    • Couldn't match type ‘f’ with ‘f0’
      ‘f’ is a rigid type variable bound by
        the type signature for:
          member :: forall k1 k2 (a :: k2) (f :: k1 -> *).
                    Typeable a =>
                    TypeRepMap f -> Bool
        at internal/Data/TypeRep/CacheMap.hs:116:1-57
      Expected type: TypeRepMap f0
        Actual type: TypeRepMap f
    • In the second argument of ‘lookup’, namely ‘trMap’
      In the expression: lookup @a trMap
      In the expression:
        case lookup @a trMap of
          Nothing -> False
          Just _ -> True
```

## Workaround 1

Change type of `member` function with explicit kind annotation:

```haskell
member :: forall k (a :: k) (f :: k -> Type) . TypeRep f -> Bool
```

I don't like this workaround because every time I call `member` function I need to do it like this:

```haskell
member @_ @Int
```

instead of just `member @Int`.

## Workaround 2 (current solution)

This solution was proposed by @int-index. It makes interface better but messes up with documentation

```haskell
type KindOf (a :: k) = k

member :: forall a (f :: KindOf a -> Type) . TypeRep f -> Bool
```

Maybe somebody knows better solution? 

If current proposal is implemented then I can just write the following (without introducing custom types):

```haskell
member :: forall {k} (a :: k) (f :: k -> Type) . TypeRep f -> Bool
```
